R/weight.R

```
@@ -59,3 +59,103 @@ setMethod("weightVariables", "VariableCatalog", function (x) {
+#' @rdname makeWeight
+#' @return
+#' A crunch Variable Definiton of the weight variable
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' mtcars$cyl <- as.factor(mtcars$cyl)
+#' mtcars$gear <- as.factor(mtcars$gear)
+#' ds <- newDataset(mtcars, "mtcars")
+#' ds$weight <- makeWeight(ds$cyl ~ c(30, 30, 40, 0), ds$gear ~ c(20, 20, 60, 0), name = "weight" )
+#' as.vector(ds$weight)
+#' }
+makeWeight <- function(..., name) {
+    expr_list <- list(...)
+    out <- list(
+        name = name,
+        derivation = list(
+            `function` = "rake",
+            args = lapply(expr_list, generateWeightEntry))
+        )
+    class(out) <- "VariableDefinition"
+    out
+}
+
+#' Generate entry for [makeWeight]
+#'
+#' Utility function to catch formula errors passed to [makeWeight], and to generate
+#' the appropriate entry.
+#'
+#' @param expr
+#' An expression
+#' @return
+#' A list of the variable id and the target weights.
+#'
+#' @keywords internal
+#' @examples
+#' \dontrun{
+#' validateWeightExpression(ds$var ~ c(10, 30, 60))
+#' }
+#'
+generateWeightEntry <- function(expr) {
+    formula <- try(as.formula(expr), silent = TRUE)
+    if (is.error(formula)) {
+        halt(dQuote(substitute(expr)), " is not a valid formula, use the form ds$var ~ c(10, 20, 30)")
+    }
+    var     <- eval(expr[[2]], environment(expr))
+    varname <- deparse(expr[[2]])
+    targets <- eval(expr[[3]])
+
+    if (!is.Categorical(var)) {
+        halt(varname, " is not a categorical crunch variable")
+    }
+
+    n_categories <- length(categories(var))
+
+    if (length(targets) > n_categories) {
+        halt("Number of targets does not match number of categories for ", varname)
+    }
+    if (!all(is.numeric(targets))) {
+        halt("Targets are not numeric for ", varname)
+    }
+    if (!(sum(targets) == 100 || sum(targets) == 1)) {
+        halt("Targets do not add up to 100% for ", varname)
+    }
+    if (sum(targets) != 1) {
+        targets <- targets / 100
+    }
+
```

Are there any other bad-input conditions you might be concerned about? What if the targets are length 0? What if there are `NA`s in `targets`?

---

Target of length 0 is going to be non-numeric so I think that one's caught. NA's produce an error, but I should probably just infer that they are zero.

---

It's technically possible though less likely for a user to accidentally produce a zero-length numeric vector than include NAs:

```
> x <- numeric(0)
> is.numeric(x)
[1] TRUE
> length(x)
[1] 0
```

And I think you should raise an error if there are NA targets because it's probably not intentional.